### DISCUSSION TITLE
Use osgDB::Registry::instance()->writeNode(...) in runtime

### DISCUSSION THREAD

**Discussion by FuXiii** [openscenegraph/OpenSceneGraph#1204] [Q&A] [answered]
1. use `a = readNodeFile("./model.ive")` read a file, then add `a` into `root` node
2. open the `viewer` to see the model in `osg`
![image](https://user-images.githubusercontent.com/40132375/217418752-29233fa9-06ad-40a1-bba0-a487d05908fb.png)
3. use `UpdateCallback` wait to write node `a` into `output.ive` file:
```CXX
//UpdateCallBack
void Update()
{
    if (this->isWirte) 
    {
        osgDB::ReaderWriter::WriteResult result = osgDB::Registry::instance()->writeNode(*a, "output.ive", osgDB::Registry::instance()->getOptions());
        if (result.success()) 
        {
            std::cout<<"writeNodeFile::Success"<<std::endl;
        }
        else
        {
            std::cout<<"writeNodeFile::Failed"<<std::endl;
        }

        this->isWirte=false;
    }
}
```
4. set `bool isWirte` to `true` and write node `a` into `output.ive`, then `Cmd` windows shows:
![image](https://user-images.githubusercontent.com/40132375/217420478-3b4abdab-d5dc-48c4-b346-b297df9f5e27.png)

5. read the  `output.ive` back into `osg`, the model become `White` model, looks like losted textures.
![image](https://user-images.githubusercontent.com/40132375/217420713-a0d9cbe8-b6cc-4cee-9ed2-c166ece17fc4.png)

---

now I had tested:
1. use `osgDB::ReaderWriter::Options::setOptionString("WirteImageHint=include")` or `osgDB::ReaderWriter::Options::setOptionString("WirteImageHint=<include>")` didn't fixed this issue.
2.  move `wirte` code from `UpdateCallback` to `Before osgViewer::Viewer::frame(...)`
```CXX
void WindowUpdateOrGameLoop()
{
        //here write ive node file before call osgViewer::Viewer::frame(...)
        wirteNodeFile();
        ...
        osgViewer::Viewer::frame(...)
} 
```
didn't fixed this issue.

---

If I call write `osgDB::Registry::instance()->writeNode(...)` before I show the `osgViewer` (maybe `osgViewer::run()`) it will write correctly. and read back `output.ive` will show correctly with right texture.

How to write correctly when osg at `Runtime`?

**Comment by robertosfield** [comment]
Could you try changing the file format your write out and read back from to .osgb rather than the very old .ive format.

Also try using osgconv to convert the files i.e.

   osgconv originalfile.ive newfile.ive
   osgconv originalfile.ive newfile.osgb

Then try running:

   osgviewer newfile.ive
   osgviewer newfile.osgb

**Comment by FuXiii** [comment, marked as answer]
I had fixed this problem by my net group friend.
After read model file into `OSG`, traverse all model's children node `stateset` an call:
```
osg::Texture2D* tex2D = dynamic_cast<osg::Texture2D>(stateset->getTextureAttribute(i, osg::StateAttribute::TEXTURE));
tex2D->setUnRefImageDataAfterApply(false);
```
And then call `osgDB::Registry::instance()->writeNode(...)` and then read the writed model file. It will draw correctly.
I don't know weather this the right method to fix this issue, if only look at the result, it maybe the right way.
